File: c++/AddOperators.cpp
class Solution {
public:
    vector<string> addOperators(string num, int target) {
        vector<string> result;
        int n = num.size();
        if (n == 0) return result;
        for(int i = 1; i <= n; i ++){
            string s = num.substr(0, i);
            long cur = stol(s);
            if (to_string(cur).size() == s.size()){ // prevent "05" convert to 5
                dfs(result, i, num, s, cur, cur, target);
            }
            
        }
        return result;
    }
    
private:
    void dfs(vector<string> &result, int start, string num, string s, long prevres, long cur, int target){
        int n = num.size();
        if (start == n && prevres == target){
            result.push_back(s);
        }
        else{
            for(int i = start + 1; i <= n; i ++){
                string t = num.substr(start, i-start);
                long now = stol(t);
                if (to_string(now).size() == t.size()){ // prevent "05" convert to 5
                    dfs(result, i, num, s+'+'+t, prevres + now, now, target);
                    dfs(result, i, num, s+'*'+t, prevres - cur + cur*now, cur*now, target);
                    dfs(result, i, num, s+'-'+t, prevres - now, -now, target);
                }
            }
        }
    }
};

File: python/addOperators.py
class Solution:
    def __init__(self):
        self.result = set()

    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        n = len(num)
        for i in range(1, n + 1):
            if i == 1 or num[0] != '0':
                self.dfs(num, i, int(num[0:i]),
                         num[0:i], int(num[0:i]), target)

        return list(self.result)

    def dfs(self, num, index, res, path, last, target):
        n = len(num)
        if index >= n:
            if res == target:
                self.result.add(path)

            return

        for i in range(1, n - index + 1):
            if i == 1 or num[index] != '0':
                cur = int(num[index:index + i])
                cur_str = num[index: index + i]
                self.dfs(num, index + i, res + cur,
                         path + '+' + cur_str, cur, target)
                self.dfs(num, index + i, res - cur, path +
                         '-' + cur_str, -cur, target)
                self.dfs(num, index + i, res - last + last * cur,
                         path + '*' + cur_str, last * cur, target)
